### PR TITLE
Expanded and clarified the Katib docs

### PR DESCRIPTION
**Pull request by sarahmaddox**
Expanded and clarified the Katib docs. Preview:

* Added [concepts](https://deploy-preview-1480--competent-brattain-de2d6d.netlify.com/docs/components/hyperparameter-tuning/overview/) in Katib, hyperparameter tuning, and neural architecture search.
* Significantly expanded the [getting-started guide](https://deploy-preview-1480--competent-brattain-de2d6d.netlify.com/docs/components/hyperparameter-tuning/hyperparameter/).
* Added a detailed guide to [configuring and running a Katib experiment](https://deploy-preview-1480--competent-brattain-de2d6d.netlify.com/docs/components/hyperparameter-tuning/experiment/).

Much of the content comes from the Katib repo on GitHub. See companion PR https://github.com/kubeflow/katib/pull/990 which moves some content out of GitHub and adds links to the website docs.

Fixes https://github.com/kubeflow/website/issues/1425

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/website/1480)

### CODE REVIEW

**Comment by sarahmaddox** [comment]
Preview:
https://deploy-preview-1480--competent-brattain-de2d6d.netlify.com/docs/components/hyperparameter-tuning/hyperparameter/

**Comment by sarahmaddox** [comment]
This PR is now ready for review. Applying a hold to allow for multiple reviewers.
/hold

**Comment by sarahmaddox** [comment]
/assign @gaocegege 
/assign @hougangliu 
/assign @richardsliu 

Please would you review this PR and its companion PR https://github.com/kubeflow/katib/pull/990. My goal is to centralize the Katib user docs, so that people can find them more easily, and to make Katib easier to use by showing people how it works within Kubeflow.

**Comment by sarahmaddox** [comment]
A note: The docs currently don't give full details of the **neural architecture search** that Katib offers. Instead, the docs introduce the feature and point to the GitHub repo for further details. At some stage, we need to document the configuration options for NAS as well as hyperparameter tuning. But that's a PR for another day.

**Comment by sarahmaddox** [comment]
Thanks @gaocegege! Yes, let's maintain the user-focused docs for Katib in this `website` repo, so that they're published on the website. We'll keep the contributor-focused docs in the `katib` repo (the *developer guide* and related docs like how to add a new algorithm). That way, we keep the docs nice and clear for people who want to use Katib for their ML workloads and don't need the contributor guides.

**Comment by sarahmaddox** [comment]
/assign @thedriftofwords 

@thedriftofwords for a tech writer review.

File: content/docs/components/hyperparameter-tuning/overview.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+The *worker job* is the process that runs to evaluate a trial and calculate
+its objective value.
+
+The worker job can be one of the following types:
+
+* [Kubernetes Job](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/)
+ (non-distributed execution). 
+* [Kubeflow TFJob](/docs/guides/components/tftraining/) (distributed execution).
+* [Kubeflow PyTorchJob](/docs/guides/components/pytorch/) (distributed
+  execution).
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/overview.md]: I've clarified that the first job type does not support distributed execution, whereas the other two types do.

File: content/docs/components/hyperparameter-tuning/hyperparameter.md [hunk 2]
```diff
@@ -136,101 +183,132 @@ Spec:
...
 ```
 
-The demo should start an experiment and run three jobs with different parameters.
-When the `spec.Status.Condition` changes to *Completed*, the experiment is 
-finished.
+When the last value in `Status.Conditions.Type` is `Succeeded`, the experiment
+is complete.
 
-### TensorFlow operator example
+View the results of the experiment in the Katib UI:
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/hyperparameter.md]: Done.

File: content/docs/components/hyperparameter-tuning/hyperparameter.md [hunk 3]
```diff
@@ -136,101 +183,132 @@ Spec:
...
 You can check the status of the experiment:
 
 ```
-kubectl -n kubeflow describe experiment pytorchjob-example
+kubectl -n kubeflow describe experiment tfjob-example
 ```
 
-## Monitoring results
+Follow the steps as described for the *random algorithm example* above, 
+to see the results of the experiment in the Katib UI.
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/hyperparameter.md]: Thanks! Done.

File: content/docs/components/hyperparameter-tuning/hyperparameter.md [hunk 4]
```diff
@@ -136,101 +183,132 @@ Spec:
...
+1. Below the graph is a list of trials that ran within the experiment:
+
+    <img src="/docs/images/katib-random-example-trials.png" 
+      alt="Trials that ran during the experiment"
+      class="mt-3 mb-3 border border-info rounded">
 
-### PyTorch example
 
-This is an example for the PyTorch operator:
+### TensorFlow operator example
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/hyperparameter.md]: Done.

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+
+This page describes in detail how to configure and run a Katib experiment.
+The experiment can perform hyperparameter tuning or a neural architecture search 
+(NAS), depending on the configuration settings.
+
+For an overview of the concepts involved, read the [introduction to 
+Katib](/docs/components/hyperparameter-tuning/overview/).
+
+<a id="docker-image"></a>
+## Packaging your training code in a container image
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Yeah, that anchor icon is a bit of a nuisance in that respect. The effect depends on the size of the screen you're using to view the doc, so I've decided not to change the text.

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+in a registry. See the [Docker
+documentation](https://docs.docker.com/develop/develop-images/baseimages/) and
+the [Kubernetes 
+documentation](https://kubernetes.io/docs/concepts/containers/images/).
+
+## Configuring the experiment
+ 
+To create a hyperparameter tuning or NAS experiment in Katib, you define the
+experiment in a YAML configuration file. The YAML file defines the range of 
+potential values (the feasible space) for the paramaters that you want to
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Done.

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+#### Grid search
+
+The algorithm name in Katib is `grid`.
+
+Grid sampling is useful when all variables are discrete (as opposed to
+continuous) and the number of possibilities is low. A grid search 
+performs an exhaustive combinatorial search over all possibilities,
+making the search process extremely long even for medium sized problems.
+
+Katib uses the [Chocolate][https://chocolate.readthedocs.io] optimization
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Thanks! Done.

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+        <td>[int]: Set <code>random_state</code> to something other than None
+          for reproducible results.</td>
+        <td>10</td>
+      </tr>
+    </tbody>
+  </table>
+</div>
+
+<a id="hyperband"></a>
+#### HYPERBAND
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Thanks, that's super useful. Done.

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+[hyperopt](http://hyperopt.github.io/hyperopt/). This method provides a 
+[forward and reverse gradient-based](https://arxiv.org/pdf/1703.01785.pdf)
+search.
+
+<a id="nas"></a>
+#### NAS using reinforcement learning
+
+The algorithm name in Katib is `nasrl`.
+
+For more information, see:
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Done.

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+  so that Katib can automatically set the values in each trial.
+
+    You can use one of the following job types to train your model:
+
+    * [Kubernetes Job](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/)
+      (non-distributed execution). 
+    * [Kubeflow TFJob](/docs/guides/components/tftraining/) (distributed
+      execution).
+    * [Kubeflow PyTorchJob](/docs/guides/components/pytorch/) (distributed
+      execution).
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Done.

**Comment by sarahmaddox** [comment]
Removing the hold now:
/hold cancel

@richardsliu Please would you review and lgtm/approve if you're happy with this update.

**Comment by sarahmaddox** [comment]
@gaocegege Can you re-LGTM this PR so I can merge it? If we need to adjust anything we can do that in followup PRs.

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+  The objective metric is also called the *target variable*. 
+  A common metric is the model's accuracy in the validation pass of the training
+  job (*validation-accuracy*). You also specify whether you want Katib to 
+  maximize or minimize the metric.
+  Katib uses the `objectiveMetricName` and `additionalMetricNames` to monitor
+  how the hyperparameters work with the model. 
+  Katib records the value of the best `objectiveMetricName` metric (maximized 
+  or minimized based on `type`) and the corresponding hyperparameter set
+  in `Experiment.status`. If the `objectiveMetricName` metric for a set of
+  hyperparameters exceeds the `goal`, Katib stops trying more hyperparameter
```
Comment (richardsliu) [inline content/docs/components/hyperparameter-tuning/experiment.md]: s/exceeds/reaches? Since the objective can be either maximize or minimize.

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 12]
```diff
@@ -0,0 +1,477 @@
...
+  A common metric is the model's accuracy in the validation pass of the training
+  job (*validation-accuracy*). You also specify whether you want Katib to 
+  maximize or minimize the metric.
+  Katib uses the `objectiveMetricName` and `additionalMetricNames` to monitor
+  how the hyperparameters work with the model. 
+  Katib records the value of the best `objectiveMetricName` metric (maximized 
+  or minimized based on `type`) and the corresponding hyperparameter set
+  in `Experiment.status`. If the `objectiveMetricName` metric for a set of
+  hyperparameters exceeds the `goal`, Katib stops trying more hyperparameter 
+  combinations. See the [EarlyStoppingSetting
```
Comment (richardsliu) [inline content/docs/components/hyperparameter-tuning/experiment.md]: Early stopping is not fully implemented yet so it should probably be removed.

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 13]
```diff
@@ -0,0 +1,477 @@
...
+    uses the [Go template format](https://golang.org/pkg/text/template/).
+    
+    You can define the job in raw string format or you can use a 
+    [ConfigMap](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/).
+
+* **parallelTrialCount**: The maximum number of hyperparameter sets that Katib
+  should train in parallel.
+
+* **maxTrialCount**: The maximum number of trials to run.
+  This is equivalent to the number of hyperparameter sets that Kabit should
```
Comment (richardsliu) [inline content/docs/components/hyperparameter-tuning/experiment.md]: Typo for Katib

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 14]
```diff
@@ -0,0 +1,477 @@
...
+  generate to test the model.
+
+* **maxFailedTrialCount**: The maximum number of failed trials before Katib
+  should stop the experiment.
+  This is equivalent to the number of failed hyperparameter sets that Kabit
```
Comment (richardsliu) [inline content/docs/components/hyperparameter-tuning/experiment.md]: Typo for Katib

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 15]
```diff
@@ -0,0 +1,477 @@
...
+  This is equivalent to the number of failed hyperparameter sets that Kabit 
+  should test.
+  If the number of failed trials exceeds `maxFailedTrialCount`, Katib stops the
+  experiment with a status of `Failed`.
+
+* **metricsCollectorSpec**: A specification of how to collect the metrics from
+  each trial, such as the accuracy and loss metrics.
+  See the [details of the metrics collector](#metrics-collector) below.
+
+* **nasConfig**: The configuration for a neural architecture search (NAS).
```
Comment (richardsliu) [inline content/docs/components/hyperparameter-tuning/experiment.md]: Should clarify here that NAS is an alpha feature.

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 16]
```diff
@@ -0,0 +1,477 @@
...
+        <td>[int]: Set <code>random_state</code> to something other than None
+          for reproducible results.</td>
+        <td>10</td>
+      </tr>
+    </tbody>
+  </table>
+</div>
+
+<a id="hyperband"></a>
+#### HYPERBAND
```
Comment (richardsliu) [inline content/docs/components/hyperparameter-tuning/experiment.md]: Why is this one in all caps?

File: content/docs/components/hyperparameter-tuning/experiment.md [hunk 17]
```diff
@@ -0,0 +1,477 @@
...
+  As an example, see the YAML file for the
+  [nasjob-example-RL](https://github.com/kubeflow/katib/blob/master/examples/v1alpha3/nasjob-example-RL.yaml).
+  The example aims to show all the possible operations. Due to the large search 
+  space, the example is not likely to generate a good result.
+
+*Background information about Katib's `Experiment` type:* In Kubernetes 
+terminology, Katib's
+[`Experiment`](https://github.com/kubeflow/katib/blob/master/pkg/apis/controller/experiments/v1alpha3/experiment_types.go#L187)
+type is a [custom resource 
+(CRD)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
```
Comment (richardsliu) [inline content/docs/components/hyperparameter-tuning/experiment.md]: Nit: CR?

File: content/docs/components/hyperparameter-tuning/experiment.md (hunk 16 above)
Comment (thedriftofwords) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Because that's how [the paper](https://arxiv.org/pdf/1603.06560.pdf) refers to it throughout its text.

Comment (richardsliu) [reply content/docs/components/hyperparameter-tuning/experiment.md]: The paper uses upper and lower-case letters for Hyperband in the title and headings. The all-caps notation is probably just formatting.

File: content/docs/components/hyperparameter-tuning/overview.md [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+
+*(To run the example that produced this graph, follow the [getting-started 
+guide](/docs/components/hyperparameter-tuning/hyperparameter/).)*
+
+Katib runs several training jobs (known as *trials*) within each
+hyperparameter tuning job (*experiment*). Each trial tests a different set of 
+hyperparameter configurations. At the end of the experiment, Katib outputs 
+the optimized values for the hyperparameters.
+
+## Neural architecture search
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/overview.md]: Thanks! I didn't realise it's alpha. I've added notes in the relevant places.

File: content/docs/components/hyperparameter-tuning/overview.md [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+### Suggestion
+
+A *suggestion* is a set of hyperparameter values that the hyperparameter
+tuning process has proposed. Katib creates a trial to evaluate the suggested
+set of values.
+
+### Trial
+
+A *trial* is one iteration of the hyperparameter tuning process. A trial
+corresponds to one worker job instance with a list of parameter assignments
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/overview.md]: Done.

File: content/docs/components/hyperparameter-tuning/overview.md [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+The list of parameter assignments corresponds to a suggestion.
+
+Each experiment runs several trials.
```
[the remark this answers is not in the document]
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/overview.md]: Done.

File: content/docs/components/hyperparameter-tuning/experiment.md (hunk 11 above)
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Good point. Done.

File: content/docs/components/hyperparameter-tuning/experiment.md (hunk 13 above)
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Done.

File: content/docs/components/hyperparameter-tuning/experiment.md (hunk 14 above)
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Done.

File: content/docs/components/hyperparameter-tuning/experiment.md (hunk 15 above)
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Done.

File: content/docs/components/hyperparameter-tuning/experiment.md (hunk 16 above)
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: I checked the paper carefully before I changed the name to caps. As far as I could see, the paper uses all caps almost always, so I decided it was a conscious branding decision. Let's go with all caps.

File: content/docs/components/hyperparameter-tuning/experiment.md (hunk 17 above)
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Done.

File: content/docs/components/hyperparameter-tuning/experiment.md (hunk 12 above)
Comment (sarahmaddox) [reply content/docs/components/hyperparameter-tuning/experiment.md]: Done.

**Comment by sarahmaddox** [comment]
Thanks for the reviews all! @richardsliu Please take another look and lgtm/approve if you're happy with the updates.
